src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,461 @@
+
+            public bool Contains(TKey item) => _dictionary.ContainsKey(item);
+
+            bool ICollection<TKey>.Remove(TKey item) =>
+                ThrowHelper.ThrowNotSupportedException_ReadOnly_Modification();
+
+            public void CopyTo(TKey[] array, int index)
+            {
+                Entry[] entries = _dictionary._entries;
+                for (int i = 0; i < _dictionary._count; i++)
+                {
+                    array[index++] = entries[i].key;
+                }
+            }
+
+            public Enumerator GetEnumerator() => new Enumerator(_dictionary); // avoid boxing
+            IEnumerator<TKey> IEnumerable<TKey>.GetEnumerator() => new Enumerator(_dictionary);
+            IEnumerator IEnumerable.GetEnumerator() => new Enumerator(_dictionary);
+
+            public struct Enumerator : IEnumerator<TKey>
+            {
+                private readonly DictionarySlim<TKey, TValue> _dictionary;
+                private int _index;
+                private TKey _current;
+
+                internal Enumerator(DictionarySlim<TKey, TValue> dictionary)
+                {
+                    _dictionary = dictionary;
+                    _index = 0;
+                    _current = default;
+                }
+
+                public TKey Current => _current;
+
+                object IEnumerator.Current => _current;
+
+                public void Dispose() { }
+
+                public bool MoveNext()
+                {
+                    if (_index == _dictionary._count)
+                    {
+                        _current = default;
+                        return false;
+                    }
+
+                    _current = _dictionary._entries[_index++].key;
+                    return true;
+                }
+
+                public void Reset()
+                {
+                    _index = 0;
+                }
+            }
+        }
+
+        public struct ValueCollection : ICollection<TValue>, IReadOnlyCollection<TValue>
+        {
+            private readonly DictionarySlim<TKey, TValue> _dictionary;
+
+            internal ValueCollection(DictionarySlim<TKey, TValue> dictionary)
+            {
+                _dictionary = dictionary;
+            }
+
+            public int Count => _dictionary._count;
+
+            bool ICollection<TValue>.IsReadOnly => true;
+
+            void ICollection<TValue>.Add(TValue item) =>
+                ThrowHelper.ThrowNotSupportedException_ReadOnly_Modification();
+
+            void ICollection<TValue>.Clear() =>
+                ThrowHelper.ThrowNotSupportedException_ReadOnly_Modification();
+
+            bool ICollection<TValue>.Contains(TValue item) =>
+                ThrowHelper.ThrowNotSupportedException(); // performance antipattern
+
+            bool ICollection<TValue>.Remove(TValue item) =>
+                ThrowHelper.ThrowNotSupportedException_ReadOnly_Modification();
+
+            public void CopyTo(TValue[] array, int index)
```

Comment:
If you make this use explicit interface implementation there won't be a need to document it. I'd imagine most uses of this would be through the interface anyway.